internal/kafka/internal/services/data_plane_kafka.go

```diff
@@ -150,30 +151,57 @@ func (d *dataPlaneKafkaService) processRealKafkaDeployment(ks *dbapi.DataPlaneKa
+	//     - FSO will not change the status of the ManagedKafka CR prior to transitioning to 'suspended' unless an error occurs.
+	//       e.g. If the Kafka instance was in a 'ready' state previously, FSO will keep reporting its status as 'ready' until it
+	//            finally reports 'suspended'. If an error occurs, the 'Ready' condition of the CR will have the values 'Status=False,Reason=Error'.
+	//  - 'resuming' state is set by the user from a 'suspending' or 'suspended' state. This must never transition to 'suspended'.
+	//  - 'suspended' state must only transition to 'resuming' or 'deprovision'.
+	//  - 'deprovision' state is set by the user. This must only transition to 'deleting' or 'failed'.
+	//     - FSO will only send the status 'deleted' if kafka was already in a 'deprovisioning' state.
+	//  - 'failed' (or 'error') state may occur at any time. However, KFM must never transition a Kafka instance to 'failed' from 'suspending' and 'suspended' states.
 	var e *serviceError.ServiceError
 	switch s := d.getManagedKafkaStatus(ks); s {
 	case statusReady:
-		// Store the routes (and create them) when Kafka is ready. By the time it is ready, the routes should definitely be there.
-		e = d.persistKafkaRoutes(kafka, ks, cluster)
-		if e == nil {
-			kafka.AdminApiServerURL = ks.AdminServerURI
-			e = d.setKafkaClusterReady(kafka)
+		if kafka.Status != constants2.KafkaRequestStatusSuspending.String() {
+			// Store the routes (and create them) when Kafka is ready. By the time it is ready, the routes should definitely be there.
+			e = d.persistKafkaRoutes(kafka, ks, cluster)
+			if e == nil {
+				kafka.AdminApiServerURL = ks.AdminServerURI
+				e = d.setKafkaClusterReady(kafka)
+			}
 		}
 	case statusInstalling:
-		// Store the routes (and create them) if they are available at this stage to lessen the length of time taken to provision the Kafka.
-		// The routes list will either be empty or complete.
-		e = d.persistKafkaRoutes(kafka, ks, cluster)
+		if kafka.Status != constants2.KafkaRequestStatusSuspending.String() {
+			// Store the routes (and create them) if they are available at this stage to lessen the length of time taken to provision the Kafka.
+			// The routes list will either be empty or complete.
+			e = d.persistKafkaRoutes(kafka, ks, cluster)
+		}
 	case statusError:
-		// when getStatus returns statusError we know that the ready
-		// condition will be there so there's no need to check for it
-		readyCondition, _ := ks.GetReadyCondition()
-		e = d.setKafkaClusterFailed(kafka, readyCondition.Message)
+		if kafka.Status != constants2.KafkaRequestStatusSuspending.String() || kafka.Status != constants2.KafkaRequestStatusSuspended.String() {
+			// when getStatus returns statusError we know that the ready
+			// condition will be there so there's no need to check for it
+			readyCondition, _ := ks.GetReadyCondition()
+			e = d.setKafkaClusterFailed(kafka, readyCondition.Message)
```

Comment:
Do we also want to set the `FailedReason` in the db when it fails during `suspending` and `suspended` state? This will be seen in the API responses, it's part of the KafkaRequest object.

Reply:
I think just logging the error is fine and we could action from there. If the error is to occur, it is not  an error due to a user action/inaction so there is nothing they could do to take the kafka from the failed state.

Reply:
Thanks @machi1990, that makes sense :)